proc_smaps.go

```diff
@@ -0,0 +1,246 @@
+// Copyright 2018 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !windows
+
+package procfs
+
+import (
+	"bufio"
+	"os"
+	"strconv"
+	"strings"
+)
+
+type ProcSMaps []*ProcSMap
+
+type ProcSMap struct {
+	*ProcMap
+	// Size of the mapping
+	Size uint64
+	// Amount of the mapping that is currently resident in RAM
+	Rss uint64
+	// Process's proportional share of this mapping
+	Pss uint64
+	// Size in bytes of clean shared pages
+	SharedClean uint64
+	// Size in bytes of dirty shared pages
+	SharedDirty uint64
+	// Size in bytes of clean private pages
+	PrivateClean uint64
+	// Size in bytes of dirty private pages
+	PrivateDirty uint64
+	// Amount of memory currently marked as referenced or accessed
+	Referenced uint64
+	// Amount of memory that does not belong to any file
+	Anonymous uint64
+	// Amount would-be-anonymous memory currently on swap
+	Swap uint64
+	// Process's proportional memory on swap
+	SwapPss uint64
+	// Page size used by the kernel to back the virtual memory area
+	KernelPageSize uint64
+	// Page size used by the processor MMU to back the virtual memory area
+	MMUPageSize uint64
+	// Kernel flags associated with the virtual memory area
+	VMFlags string
+}
+
+// SizeSum returns the sum of Pss from all mappings
+func (s ProcSMaps) SizeSum() (sum uint64) {
+	for _, x := range s {
+		sum += x.Size
+	}
+
+	return
+}
+
+// RssSum returns the sum of Pss from all mappings
+func (s ProcSMaps) RssSum() (sum uint64) {
+	for _, x := range s {
+		sum += x.Rss
+	}
+
+	return
+}
+
+// PssSum returns the sum of Pss from all mappings
+func (s ProcSMaps) PssSum() (sum uint64) {
+	for _, x := range s {
+		sum += x.Pss
+	}
+
+	return
+}
+
+// SharedCleanSum returns the sum of Pss from all mappings
```

Comment:
Looks like a copy paste typo on these comments since they all say `Pss`.  Maybe just set the comments to `...sum of all mappings` since it should be clear anyway which field is being summed.